[llm] Use new API to register custom ops for llama model

larryliu0820 opened the pull request:
Summary: Using the following 2 APIs:

* `EXECUTORCH_LIBRARY` to replace the need of a yaml file. With this macro we can directly register a custom kernel into ExecuTorch runtime.
* `WRAP_TO_ATEN` allows custom op authors to use the same kernel for ExecuTorch and PyTorch. This can be helpful during debugging.

Test Plan: Rely on the new CI job `test_llama` with `xnnpack+kv+custom` option.

Reviewers:

Subscribers:

Tasks:

Tags:

kimishpatel commented on backends/xnnpack/CMakeLists.txt:
```diff
@@ -72,6 +72,7 @@ target_include_directories(
   xnnpack_schema INTERFACE ${_xnnpack_schema__include_dir}
                            ${EXECUTORCH_ROOT}/third-party/flatbuffers/include)
 
+target_compile_options(pthreadpool PUBLIC ${_common_compile_options})
```
Why is this needed? and pthreadpool related stuff has moved to root CMakeLists

larryliu0820 replied:
my local build failed complaining no -fPIC

kimishpatel commented on examples/models/llama2/CMakeLists.txt:
```diff
@@ -114,14 +128,13 @@ if(TARGET qnn_executorch_backend)
 
 target_compile_options(llama_main PUBLIC ${_common_compile_options}
-  -DET_USE_THREADPOOL)
+                                         -DET_USE_THREADPOOL)
```
typo?

larryliu0820 replied:
Reformat cmakelists.txt

kimishpatel commented on examples/models/llama2/custom_ops/op_sdpa_aot.cpp:
```diff
@@ -0,0 +1,96 @@
+    // @lint-ignore CLANGTIDY facebook-hte-ParameterMightThrowOnCopy
+    const c10::optional<double> scale) {
+  auto output = at::empty_like(q_projected);
+  WRAP_TO_ATEN(sdpa_with_kv_cache_out_no_context, 11)
```
whats this 11?

larryliu0820 replied:
The 11th argument is `out`

kimishpatel replied:
# of args? I think there is some template magic that allows you to count number of args, right?

larryliu0820 replied:
yeah this is telling the template we need to return the 11th object, since it is `out`. See this code: https://github.com/pytorch/executorch/blob/main/extension/aten_util/make_aten_functor_from_et_functor.h#L268-L277

kimishpatel commented on examples/models/llama2/custom_ops/op_sdpa_aot.cpp:
```diff
@@ -0,0 +1,96 @@
+      "Tensor(b!) value_cache, SymInt start_pos, SymInt seq_len, Tensor? attn_mask=None, "
+      "float drpout_p=0.0, bool is_causal=False, float? scale=None) -> Tensor",
+      &torch::executor::native::sdpa_with_kv_cache_aten);
+  m.def(
```
WHy is this one needed? So that we can generate out variant one?

larryliu0820 replied:
We need both `sdpa_with_kv_cache` and `sdpa_with_kv_cache.out` in ATen so that exir is happy.

kimishpatel commented on examples/models/llama2/custom_ops/op_sdpa_with_kv_cache_test.cpp:
```diff
@@ -32,7 +32,7 @@ exec_aten::Tensor op_sdpa_with_kv_cache(
     exec_aten::Tensor& out) {
   exec_aten::RuntimeContext context{};
-  return torch::executor::llama::sdpa_with_kv_cache_outf(
+  return torch::executor::native::sdpa_with_kv_cache_out(
```
why changes the namespace?

larryliu0820 replied:
`llama` namespace was generated by `FunctionHeaderWrapper.h`. Here I added a header `op_sdpa.h` and that is using the same namespace as `op_sdpa.cpp`.

kimishpatel commented on examples/models/llama2/custom_ops/sdpa_with_kv_cache.py:
```diff
@@ -4,196 +4,21 @@
+    op = torch.ops.llama.sdpa_with_kv_cache.default
+    assert op is not None
+except:
+    # assuming we only hit this in OSS, find the default install path
+    prefix = os.environ.get("CMAKE_INSTALL_PREFIX", "../../../../cmake-out")
+    lib_path = os.path.join(prefix, "lib/libcustom_ops_aot_lib.so")
+    torch.ops.load_library(lib_path)
+    op = torch.ops.llama.sdpa_with_kv_cache.default
+    assert op is not None
```
This is a little bit clunky but I dont know how to improv e it

kimishpatel commented on examples/models/llama2/custom_ops/targets.bzl:
```diff
@@ -2,103 +2,23 @@ load("@fbsource//xplat/executorch/build:runtime_wrapper.bzl", "runtime")
-
     runtime.cxx_library(
-        name = "sdpa",
+        name = "custom_ops",
```
why rename this?

larryliu0820 replied:
A lot of places are referring to `examples/models/llama2/custom_ops:custom_ops` library name. I'm just too lazy to change all of them to `sdpa`.

kimishpatel reviewed (changes requested):
Left some comments

larryliu0820 committed:
Use new API to register custom ops for llama model (#2840)

Summary:
Using the following 2 APIs:

* `EXECUTORCH_LIBRARY` to replace the need of a yaml file. With this macro we can directly register a custom kernel into ExecuTorch runtime.
* `WRAP_TO_ATEN` allows custom op authors to use the same kernel for ExecuTorch and PyTorch. This can be helpful during debugging.


Test Plan: Rely on the new CI job `test_llama` with `xnnpack+kv+custom` option.

Reviewed By: kimishpatel

Differential Revision: D55713944

Pulled By: larryliu0820
